drivers/hm3301/hm3301.c

```diff
@@ -0,0 +1,145 @@
+#include <errno.h>
+
+#include "hm3301.h"
+#include "hm3301_constants.h"
+#include "hm3301_params.h"
+#include "timex.h"
+
+#define ENABLE_DEBUG    0
+#include "debug.h"
+
+int _set_i2c_mode(hm3301_t *dev)
+{
+    if (i2c_acquire(dev->params.i2c) != 0) {
+        DEBUG("[hm3301]: could not acquire I2C bus %d\n", dev->params.i2c);
+        return -1;
+    }
+    uint8_t cmd = HM3301_CMD_I2C_MODE;
+    int ret = i2c_write_bytes(dev->params.i2c, HM3301_I2C_ADDRESS, &cmd, 1, 0);
+    i2c_release(dev->params.i2c);
+
+    return ret;
+}
+
+int hm3301_init(hm3301_t *dev, const hm3301_params_t *params)
+{
+    assert(dev && params);
+    memset(dev, 0, sizeof(hm3301_t));
+    dev->params = *params;
+
+    if (gpio_is_valid(dev->params.reset_pin)) {
+        if (gpio_init(dev->params.reset_pin, GPIO_OUT)) {
+            DEBUG_PUTS("[hm3301]: failed to init reset pin");
+            return -EIO;
+        }
+    }
+
+    if (gpio_is_valid(dev->params.set_pin)) {
+        if (gpio_init(dev->params.set_pin, GPIO_OUT)) {
+            DEBUG_PUTS("[hm3301]: failed to init set pin");
+            return -EIO;
+        }
+        gpio_set(dev->params.set_pin);
+    }
+
+    DEBUG_PUTS("[hm3301]: reset device if reset_pin");
+    hm3301_reset(dev);
+
+    if (_set_i2c_mode(dev)) {
+        DEBUG_PUTS("[hm3301]: failed set i2c mode");
+        return -EPROTO;
+    }
+
+    return 0;
+}
+
+int hm3301_read(hm3301_t *dev, hm3301_data_t* data)
+{
+    if (i2c_acquire(dev->params.i2c) != 0) {
+        DEBUG("[hm3301]: could not acquire I2C bus %d\n", dev->params.i2c);
+        return -EPROTO;
+    }
+
+    uint8_t buf[HM3301_DATA_LENGTH] = {0};
+    if (i2c_read_bytes(dev->params.i2c, HM3301_I2C_ADDRESS, buf, HM3301_DATA_LENGTH, 0)) {
+        return -EPROTO;
+    }
+
+    /* calculate crc */
+    uint8_t crc = 0;
+    for (uint8_t i = 0; i < HM3301_DATA_LENGTH - 1; i++) {
+        crc += buf[i];
+    }
+
+    if (crc != buf[HM3301_DATA_LENGTH - 1]) {
+        DEBUG("crc mismatch expected %02x got %02x\n", buf[HM3301_DATA_LENGTH - 1], crc)
+    }
+
+    i2c_release(dev->params.i2c);
+
+    hm3301_data_t tmp = {
+        .mc_pm_1    = (buf[ 4] << 8) | buf[ 5],
+        .mc_pm_2p5  = (buf[ 6] << 8) | buf[ 7],
+        .mc_pm_10   = (buf[ 8] << 8) | buf[ 9],
+        .amc_pm_1   = (buf[10] << 8) | buf[11],
+        .amc_pm_2p5 = (buf[12] << 8) | buf[13],
+        .amc_pm_10  = (buf[14] << 8) | buf[15],
+        .nc_pm_0p3  = (buf[16] << 8) | buf[17],
+        .nc_pm_0p5  = (buf[18] << 8) | buf[19],
+        .nc_pm_1    = (buf[20] << 8) | buf[21],
+        .nc_pm_2p5  = (buf[22] << 8) | buf[23],
+        .nc_pm_5    = (buf[24] << 8) | buf[25],
+        .nc_pm_10   = (buf[26] << 8) | buf[27],
+    };
+    memcpy(data, &tmp, sizeof(hm3301_data_t));
+
+    return 0;
+}
+
+void hm3301_reset(hm3301_t *dev)
+{
+    if (gpio_is_valid(dev->params.reset_pin)) {
+        gpio_clear(dev->params.reset_pin);
+        for (uint32_t i = 0; i < 10 * (CLOCK_CORECLOCK / US_PER_SEC) ; i++) {
```

Comment:
Using `CLOCK_CORECLOCK` outside of cpu/board specific code won't work with FE310 based boards when HFROSC is used: in this case the clock speed is computed at runtime and this constant is not defined:

https://github.com/RIOT-OS/RIOT/blob/3d18f1ad672ac5eeab1f2a25b14a2aece0eb7e58/cpu/fe310/clock.c#L97-L112

This is why I thing we should not rely on a constant to get the clock speed but rather on an inline function provided by CPU implementations that would return it if possible or a computed value if not.

This was one of the indirect goals I had when I started to work on #14975.

Reply:
It doesn't really matter that much here, there is no exact time defined in the datasheet (If I recall correctly), its just about pulling the pin the pin low from some time without imposing a timer dendency. Still I added a patch so that ztimer is included if used in the build.